File: LeetCode/C/69-binary-search-x-的平方根.cpp
/*
 * @lc app=leetcode.cn id=69 lang=cpp
 *
 * [69] x 的平方根
 */

// @lc code=start
class Solution {
  public:
    int mySqrt(int x) {
        if (x < 2)
            return x;
        int l = 2, r = x >> 1;
        while (l <= r) {
            int m = (l + r) >> 1;
            if (m == x / m)
                return m;
            else if (m < x / m)
                l = m + 1;
            else
                r = m - 1;
        }
        return r;
    }
};
// @lc code=end

File: LeetCode/Go/69-binary-search-x-的平方根.go
/*
 * @lc app=leetcode.cn id=69 lang=golang
 *
 * [69] x 的平方根
 */
package leetcode

// @lc code=start
func mySqrt(x int) int {
	if x < 2 {
		return x
	}
	l, r := 2, x>>1
	for l <= r {
		m := l + (r-l)>>1
		t := m * m
		if t > x {
			r = m - 1
		} else if t < x {
			l = m + 1
		} else {
			return m
		}
	}
	return r
}

// @lc code=end
